aptos-move/framework/supra-framework/sources/automation_registry_state.move

```diff
@@ -0,0 +1,226 @@
+        registration_time: u64,
+        /// Flag indicating whether the task is active.
+        is_active: bool
+    }
+
+    #[event]
+    /// Update automation gas limit event
+    struct UpdateAutomationGasLimit has drop, store {
+        automation_gas_limit: u64
+    }
+
+    #[event]
+    /// Remove automation task registry event
+    struct RemoveAutomationTask has drop, store {
+        id: u64
+    }
+
+    public fun task_expiry_time(task: &AutomationTaskMetaData): u64 {
+        task.expiry_time
+    }
+
+    // todo : this function should call during initialzation, but since we already done genesis in that case who can access the function
+    public(friend) fun initialize(supra_framework: &signer, automation_gas_limit: u64) {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        move_to(supra_framework, AutomationRegistryState {
+            tasks: enumerable_map::new_map(),
+            current_index: 0,
+            gas_committed_for_next_epoch: 0,
+            automation_gas_limit
+        })
+    }
+    public(friend) fun on_new_epoch(epoch_interval: u64) acquires AutomationRegistryState {
+        let state = borrow_global_mut<AutomationRegistryState>(@supra_framework);
+        let ids = enumerable_map::get_map_list(&state.tasks);
+
+        let current_time = timestamp::now_seconds();
+        let expired_task_gas = 0;
+
+        // Perform clean up and updation of state
+        vector::for_each(ids, |id| {
+            let task = enumerable_map::get_value_mut(&mut state.tasks, id);
+
+            // Tasks that are active during this new epoch but will be already expired for the next epoch
+            if (task.expiry_time <= (current_time + epoch_interval)) {
+                expired_task_gas = expired_task_gas + task.max_gas_amount;
+            };
+
+            if (task.expiry_time <= current_time) {
+                enumerable_map::remove_value(&mut state.tasks, id);
+            } else {
+                task.is_active = true;
+            }
+        });
+
+        // Adjust the gas committed for the next epoch by subtracting the gas amount of the expired task
+        state.gas_committed_for_next_epoch = state.gas_committed_for_next_epoch - expired_task_gas;
+    }
+
+
+    /// Registers a new automation task entry.
+    public(friend) fun register(
+        owner: &signer,
+        payload_tx: vector<u8>,
+        expiry_time: u64,
+        max_gas_amount: u64,
+        gas_price_cap: u64,
+        registration_epoch: u64,
+        tx_hash: vector<u8>,
+    ) acquires AutomationRegistryState {
+        let registry_data = borrow_global_mut<AutomationRegistryState>(@supra_framework);
+
+        let committed_gas = registry_data.gas_committed_for_next_epoch + max_gas_amount;
+        assert!(committed_gas < registry_data.automation_gas_limit, EGAS_AMOUNT_UPPER);
+        registry_data.gas_committed_for_next_epoch = committed_gas;
+        let task_index = registry_data.current_index;
```

Comment:
If we are not incrementing `task_index` here, does that mean the task registration ID will start from 0?
This is because we initialized it with `current_index: 0`.

Reply:
usually in this world indexing starts from 0, if we want/prefer otherwise I will revert this change.

Reply:
I just wanted to confirm that in the initial version, I started from 1, but It's okay with starting from 0 instead.